spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraStrategies.scala

```diff
@@ -121,7 +121,8 @@ private[cassandra] trait CassandraStrategies {
         val eqPredicates = eqColumns.flatMap(eqPredicatesByName)
         val optionalNonEqPredicate = for {
           c <- otherColumns.headOption.toSeq
-          p <- firstNonEmptySeq(rangePredicatesByName(c), inPredicatesByName(c).filter(_ => eqColumns.nonEmpty))
+          p <- firstNonEmptySeq(rangePredicatesByName(c), inPredicatesByName(c).filter(
+            _ => eqColumns.nonEmpty && c==clusteringColumns.last))
```

Comment:
c <- otherColumns.headOption.toSeq only returns the first other column, so all previous columns have Eq predicates if there eqColumns are not empty.

Only the last cluster column are allowed to have IN predicate. I will add some more tests.